## An image-based guide of Fibromine</strong>

#### 1. How to explore a single dataset using the <i class="fas fa-database"></i> Dataset explorer?

##### 1.1 Explore one - or integrate multiple - transcriptomics datasets 

1. Go to **Dataset explorer --> Transcriptomic datasets --> Datasets tab** and **choose** (via clicking) the dataset/comparison of interest from the displayed table as shown below and then **press the Search button (1)**. Every column of the table can be dynamically filtered. 

	<a href= "howToScreenshots/datasetExpl1_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/datasetExpl1_mod.png" alt="image" style="width:900px;height:500px" class="center">
	</a>

2. To interrogate the results navigate to the **Dataset explorer --> Transcriptomic datasets --> DEA statistics** tab **(1)**. If more than one datasets have been selected then consensus differentially expressed genes are presented at the **Transcriptomics summary** tab **(2)**. In the case of a single dataset interrogation the aforementioned tab displays the same data presented at the **Transcriptomics analytically** tab **(4)**, namely all the DEA results of the dataset. Finally, **Proteomics summary** tab **(3)** presents the differentially expressed proteins coded by the genes of the *Transcriptomics summary* tab, if any.

	<a href= "howToScreenshots/datasetExpl2_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/datasetExpl2_mod.png" alt="image" style="width:900px;height:500px" class="center"/>
	</a>

	<a href= "howToScreenshots/datasetExpl3_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/datasetExpl3_mod.png" alt="image" style="width:900px;height:500px" class="center"/>
	</a>

	<a href= "howToScreenshots/datasetExpl4_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/datasetExpl4_mod.png" alt="image" style="width:900px;height:500px" class="center"/>
	</a>

3. **Concerning dataset(-s)** tab (1) gives to the user the ability to plot an interactive heatmap and volcano plot for any of the datasets queried in step 1. **To plot** any of the two, **select a dataset** from the tab's table (2) and then **press the respective _Plot_ button**, **(3-4)**.

	<a href= "howToScreenshots/datasetExpl5_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/datasetExpl5_mod.png" alt="image" style="width:900px;height:700px" class="center"/>
	</a>

##### 1.2 Explore one - or integrate multiple - proteomics dataset
	
1. Go to **Dataset explorer --> Proteomic datasets --> Datasets tab** and **choose** (via clicking) the dataset/comparison of interest from the displayed table as shown below and then **press the _Search_ button (1)**. Every column of the table can be dynamically filtered. 

	<a href= "howToScreenshots/datasetExpl6_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/datasetExpl6_mod.png" alt="image" style="width:900px;height:600px" class="center">
	</a>

2. To interrogate the results navigate to the **Dataset explorer --> Proteomic datasets --> DEA data** tab **(1)**. If more than one datasets have been selected then consensus differentially expressed genes are presented at the **Proteomics summary** tab **(2)**. In the case of a single dataset interrogation the aforementioned tab displays the same data presented at the **Proteomics analytically** tab **(3)**, namely all the DEA results of the dataset.

	<a href= "howToScreenshots/datasetExpl7_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/datasetExpl7_mod.png" alt="image" style="width:900px;height:500px" class="center"/>
	</a>

	<a href= "howToScreenshots/datasetExpl8_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/datasetExpl8_mod.png" alt="image" style="width:900px;height:500px" class="center"/>
	</a>

---

#### 2. How to search for the expression pattern of a specific (or multiple) gene?

1. Navigate to the **Gene explorer** and type your gene(-s) of interest into the **search box (1)**. Then select in which of the supported species you want to search using the **radio buttons (2)** below the search box and finally press the **Search button (3)**. The results presented into the following screenshots can be replicated using the **Example button (4)**. For more details please hover over any of the aforementioned buttons/boxes or even press on the **About [...]** buttons of the tab.

	<a href= "howToScreenshots/geneExpl1_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/geneExpl1_mod.png" alt="image" style="width:900px;height:1000px" class="center"/>
	</a>

2. The results of the respective query are constructed as thus: (i) the **Queried genes (4)** box displays some general information regarding the queried genes, (ii) differential expression statistics for the gene(-s) queried are visualized into the **Expression data tab --> DEG statistics table (5)**, (iii) if there are any differentially expressed proteins (in the proteomics datasets) coded by the queried genes, these are presented at the **Expression data tab --> DEP statistics table (6)** at the bottom of the page, (iv) Gene ontology terms associated with the gene(-s) of interest are presented at the **Gene ontology tab (7)** and finally, (v) any RefSeq related sequences or miRNA-mRNA potential interactions as described in miRDB are presented at the **RefSeq - miRNA tab (8)**.
	
	<a href= "howToScreenshots/geneExpl2_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/geneExpl2_mod.png" alt="image" style="width:900px;height:700px" class="center"/>
	</a>
	<a href= "howToScreenshots/geneExpl3_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/geneExpl3_mod.png" alt="image" style="width:900px;height:700px" class="center"/>
	</a>

---

#### 3. Search for the expression pattern of a specific protein.

1. Navigate to the **Protein explorer**, type the name of the gene coding your protein of interest into the **search box (1)** at the upper left corner of the explorer and then press the **Search (2)** button below. The results presented are constructed as thus: (i) the **info boxes (3)** at the top of the page summarise some important information regarding the protein queried, while (ii) the **Concerning protein (4)** tab displays some general information sourced from UniProt/SwissProt. *Note: the results presented in the following example screenshots can be recreated by pressing the Example button below the Search button of the explorer.*

	<a href= "howToScreenshots/proteinExpl1_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/proteinExpl1_mod.png" alt="image" style="width:900px;height:500px" class="center"/>
	</a>

2. Finally the differential expression data, if any, are presented at the **Differential expression (5)** tab.

	<a href= "howToScreenshots/proteinExpl2_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/proteinExpl2_mod.png" alt="image" style="width:900px;height:300px" class="center"/>
	</a>

---

#### 4. Creation of a condition-specific protein-protein interaction network.

1. To construct a condition specific protein-protein interaction (PPI) network, first search for a specific protein as described in step 1 of **3. Search for the expression pattern of a specific protein.**.

2. Then, move to the **PPI network (1)** tab of the **Protein explorer** and press the **Plot (2)** button. The network should be created within a few seconds.

	<a href= "howToScreenshots/proteinExpl3_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/proteinExpl3_mod.png" alt="image" style="width:900px;height:600px" class="center"/>
	</a>

3. To annotate the network **select tissue (3)** and then **select comparison (4)** from the homonym dropdown menus at the right of the network. Finally press the **Annotate (5)** button and wait for a few seconds. In order to annotate the network using data from different conditions, press the **Reset** button next to the **Annotation** one.

	<a href= "howToScreenshots/proteinExpl4_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/proteinExpl4_mod.png" alt="image" style="width:900px;height:600px" class="center"/>
	</a>

4. For transparency reasons, the data used to annotate the network are presented into the **Data used (6)** tab.
	
	<a href= "howToScreenshots/proteinExpl5_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/proteinExpl5_mod.png" alt="image" style="width:900px;height:500px" class="center"/>
	</a>

---

#### 5. Download data

1. To download data for any downstream analysis/procedure, navigate to the **Download data** tab and use the **dropdown menus (1-2)** of the page to fetch the normalized data of any transcriptomic dataset, or the differential expression data of any proteomic one.

	<a href= "howToScreenshots/download_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/download_mod.png" alt="image" style="width:900px;height:300px" class="center"/>
	</a>

---

#### 6. Get more detailed information

1. To obtain more detailed information about Fibromine, please visit the **Docs** tab. 

	<a href= "howToScreenshots/doc_mod.png" target="_blank" rel='noopener noreferrer'> 
		<img src= "./howToScreenshots/doc_mod.png" alt="image" style="width:900px;height:400px" class="center"/>
	</a>